From 956430424a57c6b4b359ea92841464285377e232 Mon Sep 17 00:00:00 2001
From: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
Date: Thu, 29 Oct 2020 19:25:01 +0700
Subject: [PATCH 340/442] drm: rcar-du: Fix hot plug detection method

If the resource information on ADV7511 HDMI transmitter of DT
has registration of GPIO Pin, hot plug is detected by GPIO
external interrupt.
If there is no registration of GPIO Pin, hot plug is detected
by polling as usual.

Signed-off-by: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
Signed-off-by: Hung Dong <hung.dong.xd@rvc.renesas.com>
Signed-off-by: Duy Dang <duy.dang.yb@renesas.com>
---
 drivers/gpu/drm/rcar-du/rcar_du_hdmicon.c | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/rcar-du/rcar_du_hdmicon.c b/drivers/gpu/drm/rcar-du/rcar_du_hdmicon.c
index 6c92714..0d36e82 100644
--- a/drivers/gpu/drm/rcar-du/rcar_du_hdmicon.c
+++ b/drivers/gpu/drm/rcar-du/rcar_du_hdmicon.c
@@ -16,6 +16,8 @@
 #include <drm/drm_crtc.h>
 #include <drm/drm_crtc_helper.h>
 #include <drm/drm_encoder_slave.h>
+#include <linux/gpio.h>
+#include <linux/of_gpio.h>
 
 #include "rcar_du_drv.h"
 #include "rcar_du_encoder.h"
@@ -85,6 +87,7 @@ int rcar_du_hdmi_connector_init(struct rcar_du_device *rcdu,
 	struct rcar_du_connector *rcon;
 	struct drm_connector *connector;
 	int ret;
+	struct device_node *np;
 
 	rcon = devm_kzalloc(rcdu->dev, sizeof(*rcon), GFP_KERNEL);
 	if (rcon == NULL)
@@ -94,7 +97,12 @@ int rcar_du_hdmi_connector_init(struct rcar_du_device *rcdu,
 	connector->display_info.width_mm = 0;
 	connector->display_info.height_mm = 0;
 	connector->interlace_allowed = true;
-	connector->polled = DRM_CONNECTOR_POLL_HPD;
+	np = of_find_node_by_name(NULL, "adv7511w");
+	if ((np) && (of_get_gpio(np, 0) > 0))
+		connector->polled = DRM_CONNECTOR_POLL_HPD;
+	else
+		connector->polled = DRM_CONNECTOR_POLL_CONNECT |
+							DRM_CONNECTOR_POLL_DISCONNECT;
 
 	ret = drm_connector_init(rcdu->ddev, connector, &connector_funcs,
 				 DRM_MODE_CONNECTOR_HDMIA);
-- 
2.7.4

